### PR TITLE
USHIFT-6912: Add DNS deployment resource configuration support

### PR DESCRIPTION
**Pull request by Neilhamza**
Allow users to configure CPU and memory resources for the dns container in the dns-default DaemonSet via dns.resources in config.yaml. Defaults to the current hardcoded values (cpu=50m, memory=70Mi requests, no limits) when not configured.

## Summary by CodeRabbit

* **New Features**
  * DNS container CPU/memory can be configured via requests and optional limits; defaults apply when requests omitted.

* **Documentation**
  * Configuration examples and default settings updated to show dns.resources.requests and dns.resources.limits.

* **Bug Fixes**
  * Validation now rejects invalid quantities and limits smaller than requests; invalid configs prevent startup with clear errors.

* **Tests**
  * Added unit and integration tests plus a Robot Framework suite covering parsing, validation, overrides, and runtime behavior.

### CODE REVIEW

**Comment by Neilhamza** [comment]
@coderabbitai resolve

**Comment by agullon** [comment]
hi @Neilhamza thank you very much for creating the PR early in the dev cycle.
Only one question from QE POV, are you planning to add RF tests in current PR or in a separate one?
I think best practise is to added them also in this PR to validate changes are working as expected. Let me know if you have any questions or I can help you with RF tests.

**Comment by Neilhamza** [comment]
Thanks for the feedback @agullon !! i was thinking about a seperate PR so i wont bloat this one, but if adding the tests here is the accepted approach i can do it, what do you think? :)

**Comment by agullon** [comment]
> Thanks for the feedback @agullon !! i was thinking about a seperate PR so i wont bloat this one, but if adding the tests here is the accepted approach i can do it, what do you think? :)

I don't have a strong opinion about when/where add the RF, it's up to you. As I said I think it's better to add them along with product change. This PR is not that big, no problem.

**Comment by Neilhamza** [comment]
@coderabbitai perform full review

**Comment by Neilhamza** [comment]
@agullon added RF testscoverage - : default resources, custom requests+limits, requests only, partial requests, invalid quantity rejection, and  DNS resolution after change

File: assets/components/openshift-dns/dns/daemonset.yaml [hunk 1]
```diff
@@ -57,8 +57,17 @@ spec:
...
+              memory: {{ .DNSMemoryRequest }}
+            {{- if .DNSHasLimits }}
+            limits:
+              {{- if .DNSCPULimit }}
+              cpu: {{ .DNSCPULimit }}
+              {{- end }}
+              {{- if .DNSMemoryLimit }}
+              memory: {{ .DNSMemoryLimit }}
+              {{- end }}
+            {{- end }}
```
Comment (pmtk) [inline assets/components/openshift-dns/dns/daemonset.yaml]: Why not just give customers whole `resources: {}`? If they define anything, we take it whatever they provide. Would simplify templating greatly

Comment (eslutsky) [reply assets/components/openshift-dns/dns/daemonset.yaml]: I agree . since resource is something  generic we might be able to reuse it for different components besides dns. (ie ingress)

**Comment by pacevedom** [comment]
/hold
Wait for [enhancement](https://github.com/openshift/enhancements/pull/1989) to merge first to avoid subsequent PRs.

File: assets/components/openshift-dns/dns/daemonset.yaml (hunk 1 above)
Comment (Neilhamza) [reply assets/components/openshift-dns/dns/daemonset.yaml]: Nice! simplified the template to pass resource map directly with range loops of 2 params instead of 7 and is reusable for other components

File: test/suites/configuration1/dns-resource-configuration.robot [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+...                                 \ \ resources:
+...                                 \ \ \ requests:
+...                                 \ \ \ \ cpu: "200m"
+...                                 \ \ \ limits:
+...                                 \ \ \ \ cpu: "50m"
+
+
+*** Test Cases ***
+Default DNS Resources
+    [Documentation]    Verify default DNS resources when no custom config is applied
```
Comment (agullon) [inline test/suites/configuration1/dns-resource-configuration.robot:68]:
```suggestion
    [Documentation]    Verify default DNS resources when no custom config is applied
    [Setup]    Remove DNS Resource Config
```
Tests must be indempotent, meaning regardless the previous status of the system they must always pass. For that reason the test must enforce default DNS config before checking it.

File: test/suites/configuration1/dns-resource-configuration.robot [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+    ${limits}=    Oc Get JsonPath    daemonset    openshift-dns    dns-default
+    ...    .spec.template.spec.containers[0].resources.limits
+    Should Be Equal As Strings    ${cpu}    100m
+    Should Be Equal As Strings    ${memory}    150Mi
+    Should Be Empty    ${limits}
+    [Teardown]    Remove DNS Resource Config
+
+Partial Requests Preserves Defaults
+    [Documentation]    Configure only CPU request and verify memory default is preserved
+    [Setup]    Apply DNS Resource Config    ${DNS_PARTIAL_REQUESTS}
```
Comment (agullon) [inline test/suites/configuration1/dns-resource-configuration.robot:101]: the setup of this test only set requests.cpuin the config.
What would happen if requests.memory was already set to a value different than 70Mi (default) before this test starts? I think it would failed because the different value will still be present when the check is done.
To avoid this test must remove all configurations and then apply any config before doing any check.

Something not great about this suggestion is that MicroShift will restart many times (I think 3) on every test, which will lead into long test executions. Checks take less than a second. Successful microshift restart at least 1 min. We must have this in mind. An alternative may be running all the tests in a predefined order to reduce the number of times microshift restarts.

File: test/suites/configuration/dns-resource-configuration.robot [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+...                                 \ \ \ requests:
+...                                 \ \ \ \ cpu: "200m"
+...                                 \ \ \ limits:
+...                                 \ \ \ \ cpu: "50m"
+
+
+*** Test Cases ***
+Default DNS Resources
+    [Documentation]    Verify default DNS resources when no custom config is applied
+    ${cpu}=    Oc Get JsonPath    daemonset    openshift-dns    dns-default
```
Comment (agullon) [inline test/suites/configuration/dns-resource-configuration.robot]: `Oc Get JsonPath    daemonset    openshift-dns    dns-default` is duplicated 15 times. Can you move it into a Keyword

File: test/suites/configuration/dns-resource-configuration.robot [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+    ...    .spec.template.spec.containers[0].resources.requests.cpu
+    ${memory}=    Oc Get JsonPath    daemonset    openshift-dns    dns-default
+    ...    .spec.template.spec.containers[0].resources.requests.memory
```
Comment (agullon) [inline test/suites/configuration/dns-resource-configuration.robot]: You may also want to extract `.spec.template.spec.containers[0].resources` into a var `${DNS_RESOURCE_PATH}    .spec.template.spec.containers[0].resources` and then use it `${DNS_RESOURCE_PATH}.requests.cpu`

File: test/suites/configuration1/dns-resource-configuration.robot (hunk 2 above)
Comment (Neilhamza) [reply test/suites/configuration1/dns-resource-configuration.robot:68]: Thanks for the detailed review @agullon , i committed this suggested change

File: test/suites/configuration1/dns-resource-configuration.robot (hunk 3 above)
Comment (Neilhamza) [reply test/suites/configuration1/dns-resource-configuration.robot:101]: fixed - now we remove any existing drop-in before applying the new one

File: test/suites/configuration/dns-resource-configuration.robot (hunk 4 above)
Comment (Neilhamza) [reply test/suites/configuration/dns-resource-configuration.robot]: done created  Get DNS Resource Value keyword that wraps it

File: test/suites/configuration/dns-resource-configuration.robot (hunk 5 above)
Comment (Neilhamza) [reply test/suites/configuration/dns-resource-configuration.robot]: fixed

File: test/suites/configuration/dns-resource-configuration.robot [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+    [Documentation]    Get a resource value from the dns-default DaemonSet
+    [Arguments]    ${jsonpath}
+    ${value}=    Oc Get JsonPath    daemonset    openshift-dns    dns-default    ${jsonpath}
+    RETURN    ${value}
+
+Apply DNS Resource Config
+    [Documentation]    Remove any existing drop-in, apply a new DNS resource config and restart MicroShift
+    [Arguments]    ${config}
+    Remove Drop In MicroShift Config    ${DNS_DROPIN}
+    Drop In MicroShift Config    ${config}    ${DNS_DROPIN}
```
Comment (agullon) [inline test/suites/configuration/dns-resource-configuration.robot]: Inside `Drop In MicroShift Config` MicroShift config is being removed.
By only adding `Remove Drop In MicroShift Config` before `Drop In MicroShift Config` you are not changing any MicroShift config, only some files. This is because MicroShift config is only update it when microshift restarts. MicroShift must always restart after any config change.


```suggestion
    Remove DNS Resource Config
    Drop In MicroShift Config    ${config}    ${DNS_DROPIN}
```

File: test/suites/configuration/dns-resource-configuration.robot [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+    Restart MicroShift
+
+Apply Invalid DNS Resource Config
+    [Documentation]    Apply an invalid DNS resource config that should prevent MicroShift from starting
+    [Arguments]    ${config}
+    Remove Drop In MicroShift Config    ${DNS_DROPIN}
```
Comment (agullon) [inline test/suites/configuration/dns-resource-configuration.robot]:
```suggestion
    Remove DNS Resource Config
```

File: test/suites/configuration/dns-resource-configuration.robot [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+
+Remove DNS Resource Config
+    [Documentation]    Remove the DNS resource drop-in config and restart MicroShift
+    Remove Drop In MicroShift Config    ${DNS_DROPIN}
+    Restart MicroShift
+
+Restore Valid DNS Config
+    [Documentation]    Remove invalid config and restore MicroShift to a working state
+    Remove Drop In MicroShift Config    ${DNS_DROPIN}
+    Restart MicroShift
```
Comment (agullon) [inline test/suites/configuration/dns-resource-configuration.robot]: These 2 are the same, let's use only one.

Comment (Neilhamza) [reply test/suites/configuration/dns-resource-configuration.robot]: fixed

**Comment by agullon** [comment]
it looks good to me from QE point of view
great work @Neilhamza !!!

File: pkg/config/dns.go [hunk 9]
```diff
@@ -13,6 +15,20 @@ const (
...
+type DNSResources struct {
+	// Requests specifies the minimum resources required for the dns container.
+	// Valid keys are "cpu" and "memory". Values must be valid Kubernetes resource quantities.
+	// When not set, defaults to cpu=50m, memory=70Mi.
+	Requests map[string]string `json:"requests,omitempty"`
+
+	// Limits specifies the maximum resources the dns container can use.
+	// Valid keys are "cpu" and "memory". Values must be valid Kubernetes resource quantities.
+	// When not set, no limits are applied.
+	Limits map[string]string `json:"limits,omitempty"`
```
Comment (pacevedom) [inline pkg/config/dns.go:29]: Is there any way to specify the defaults here? So that the config.yaml and the openapi spec have them too.

File: test/suites/configuration1/dns-resource-configuration.robot [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+    Restart MicroShift
+
+Apply Invalid DNS Resource Config
+    [Documentation]    Apply an invalid DNS resource config that should prevent MicroShift from starting
+    [Arguments]    ${config}
+    Remove DNS Resource Config
+    Drop In MicroShift Config    ${config}    ${DNS_DROPIN}
+    ${cursor}=    Get Journal Cursor
+    VAR    ${CURSOR}=    ${cursor}    scope=TEST
+    Run Keyword And Expect Error    0 != 1    Restart MicroShift
```
Comment (pacevedom) [inline test/suites/configuration1/dns-resource-configuration.robot:174]: Maybe use `Keyword Should Fail`?

File: test/suites/configuration1/dns-resource-configuration.robot [hunk 11]
```diff
@@ -0,0 +1,179 @@
...
+
+Requests Only Without Limits
+    [Documentation]    Configure only requests without limits and verify no limits are injected
+    [Setup]    Apply DNS Resource Config    ${DNS_REQUESTS_ONLY}
+    ${cpu}=    Get DNS Resource Value    ${DNS_RESOURCE_PATH}.requests.cpu
+    ${memory}=    Get DNS Resource Value    ${DNS_RESOURCE_PATH}.requests.memory
+    ${limits}=    Get DNS Resource Value    ${DNS_RESOURCE_PATH}.limits
+    Should Be Equal As Strings    ${cpu}    100m
+    Should Be Equal As Strings    ${memory}    150Mi
+    Should Be Empty    ${limits}
```
Comment (pacevedom) [inline test/suites/configuration1/dns-resource-configuration.robot:96]: Using limits as an empty map would bypass the check?

File: pkg/config/dns.go [hunk 12]
```diff
@@ -90,14 +118,62 @@ func (t *DNS) validateHosts() error {
...
+			return fmt.Errorf("invalid dns resource limit %s=%q: %v", key, val, err)
+		}
+	}
+	for key, limitVal := range t.Resources.Limits {
+		reqVal, ok := t.Resources.Requests[key]
+		if !ok {
+			continue
+		}
+		limit := resource.MustParse(limitVal)
+		req := resource.MustParse(reqVal)
```
Comment (pacevedom) [inline pkg/config/dns.go:172]: These could be constants?

File: pkg/config/config.go [hunk 13]
```diff
@@ -442,6 +442,24 @@ func (c *Config) incorporateUserSettings(u *Config) {
 			c.DNS.Hosts.File = u.DNS.Hosts.File
 		}
 	}
+
+	// DNS resource configuration - merge key-by-key to preserve defaults
+	if u.DNS.Resources.Requests != nil {
+		if c.DNS.Resources.Requests == nil {
+			c.DNS.Resources.Requests = make(map[string]string)
+		}
```
Comment (pacevedom) [inline pkg/config/config.go:452]: This is correctly handled, but is there a test for this case?

File: pkg/config/dns.go (hunk 9 above)
Comment (Neilhamza) [reply pkg/config/dns.go:29]: i would need to modify  cmd/generate-config/parser.go to handle object defaults i'm not sure if that should be a scope of this PR - what do you think?

File: test/suites/configuration1/dns-resource-configuration.robot (hunk 10 above)
Comment (Neilhamza) [reply test/suites/configuration1/dns-resource-configuration.robot:174]: Run Keyword And Expect Error is the existing pattern used in tls-configuration.robot and configuration.robot. Keyword Should Fail isn't used anywhere in the test suite.  should i keep the existing pattern for consistency? or you would prefer to use Should Fail anyway?

File: pkg/config/dns.go (hunk 9 above)
Comment (Neilhamza) [reply pkg/config/dns.go:29]: The kubebuilder annotation +kubebuilder:default={"cpu":"50m","memory":"70Mi"} adds the default to the OpenAPI spec, but the config generator (cmd/generate-config/parser.go)  doesn't render map-type defaults

File: test/suites/configuration1/dns-resource-configuration.robot (hunk 11 above)
Comment (Neilhamza) [reply test/suites/configuration1/dns-resource-configuration.robot:96]: empty limits map (limits: {}) just means "no limits"  same as not setting it. The jsonpath returns empty, and "Should Be Empty" passes :dancers:

File: pkg/config/dns.go (hunk 12 above)
Comment (Neilhamza) [reply pkg/config/dns.go:172]: limitVal and reqVal arent' costant values they come from user config and change each loop iteration so MustParse is safe here

File: pkg/config/config.go (hunk 13 above)
Comment (Neilhamza) [reply pkg/config/config.go:452]: Added dns-resources-limits-only test case

File: pkg/config/dns.go (hunk 9 above)
Comment (pacevedom) [reply pkg/config/dns.go:29]: Well , fillDefaults will do then. Thanks!

File: test/suites/configuration1/dns-resource-configuration.robot (hunk 10 above)
Comment (pacevedom) [reply test/suites/configuration1/dns-resource-configuration.robot:174]: Was asking because if the error changes this is brittle, but if thats the pattern lets address it in a different change.

File: test/suites/configuration1/dns-resource-configuration.robot (hunk 11 above)
Comment (pacevedom) [reply test/suites/configuration1/dns-resource-configuration.robot:96]: Thanks!

File: pkg/config/dns.go (hunk 12 above)
Comment (pacevedom) [reply pkg/config/dns.go:172]: Oh wait, I meant the constants in dnsMinimumRequests, but that might be too convoluted because of types.

File: pkg/config/config.go (hunk 13 above)
Comment (pacevedom) [reply pkg/config/config.go:452]: Thanks

File: pkg/config/config_test.go [hunk 14]
```diff
@@ -904,6 +979,113 @@ func TestValidate(t *testing.T) {
 			}(),
 			expectErr: true,
 		},
+		{
```
Comment (pacevedom) [inline pkg/config/config_test.go:982]: Missing tests with 0 and negative quantities.

File: test/suites/configuration1/dns-resource-configuration.robot
Comment (pacevedom) [inline test/suites/configuration1/dns-resource-configuration.robot:1]: Keep in mind the directory is now `test/suites/configuration1`

**Review by pacevedom — COMMENTED** [review]
/hold cancel

Minor nits, looking good overall!

**Review by pacevedom — APPROVED** [review]
/lgtm
